tachyon/crypto/sumcheck/sumcheck_verifier.h

```diff
@@ -0,0 +1,267 @@
+#include "tachyon/crypto/sumcheck/sumcheck_verifying_key.h"
+
+namespace tachyon::crypto {
+
+template <typename F>
+F InterpolateUniPoly(const std::vector<F>& poly, const F& evaluation_point);
+
+// Subclaim created when verifier is convinced.
+template <typename F>
+struct Subclaim {
+  // The multi-dimensional point this multilinear evaluation is evaluated at.
+  std::vector<F> point;
+  // The expected evaluation.
+  F expected_evaluation;
+};
+
+template <typename MLE>
+class SumcheckVerifier {
+ public:
+  using F = typename MLE::Field;
+
+  SumcheckVerifier() = default;
+  explicit SumcheckVerifier(const SumcheckVerifyingKey& key)
+      : num_variables_(key.num_variables),
+        max_evaluations_(key.max_evaluations) {
+    randomness_.reserve(key.num_variables);
+    polynomials_received_.reserve(key.num_variables);
+  }
+  explicit SumcheckVerifier(SumcheckVerifyingKey&& key)
+      : num_variables_(key.num_variables),
+        max_evaluations_(key.max_evaluations) {
+    randomness_.reserve(key.num_variables);
+    polynomials_received_.reserve(key.num_variables);
+  }
+
+  // Runs a "round" given a |SumcheckProverMsg|, sampling and storing a random
+  // value.
+  //
+  // While a true verifier round should perform actual verification, |Round()|
+  // merely samples and stores random values. Verifications will be performed
+  // altogether in |CheckAndGenerateSubclaim| after all prover and verifier
+  // sub-rounds are subsequently finished.
+  //
+  // Referencing https://people.cs.georgetown.edu/jthaler/sumcheck.pdf
+  // |random_value| is the challenge that will given from the Verifier
+  // to the Prover aka rᵢ.
+  template <size_t MaxDegree>
+  SumcheckVerifierMsg<F> Round(SumcheckProverMsg<F, MaxDegree>&& prover_msg) {
+    CHECK_LE(round_, num_variables_);
+
+    const F random_value = F::Random();
+    randomness_.push_back(random_value);
+    polynomials_received_.push_back(
+        std::move(prover_msg.evaluations.evaluations()));
+
+    ++round_;
+
+    return {std::move(random_value)};
+  }
+
+  // Verifies the sumcheck phases' validity and generates a subclaim.
+  //
+  // If the asserted sum is correct, then the multilinear polynomial evaluated
+  // at |subclaim.point| is |subclaim.expected_evaluation|. Otherwise, it is
+  // highly unlikely that these two will be equal. A larger field size
+  // guarantees a smaller soundness error.
+  //
+  // Referencing https://people.cs.georgetown.edu/jthaler/sumcheck.pdf
+  // |CheckAndGenerateSubclaim()| verifies for each round i:
+  //   gᵢ(0) + gᵢ(1) =  gᵢ₋₁(rᵢ₋₁)
+  // Note: For the first round, gᵢ₋₁(rᵢ₋₁) = H = |LinearCombination.Combine()|
+  //   Meanwhile, for the last round, gᵢ₋₁(rᵢ₋₁) = the |LinearCombination|
+  //   evaluated on the multivariate point of Verifier's challenges.
+  bool CheckAndGenerateSubclaim(const F& asserted_sum, Subclaim<F>& subclaim) {
+    // Verifier subrounds have not been completed.
+    CHECK_EQ(round_ - 1, num_variables_);
```

Comment:
You need to add `round_` by 1, right?

```suggestion
    CHECK_EQ(round_++, num_variables_);
```

Reply:
this function should run after all prover/sub-verifier rounds -> at this point, `round_` is `num_variables +1` given the logic in the code. This check makes sure that all the prover/sub-verifier rounds have finished before we start truly verifying everything.

Reply:
Then to prevent it from being called again, how about changing like below?

```suggestion
    CHECK_EQ(round_++, num_variables_ + 1);
```